### PR TITLE
Twice differentiability of pointwise functions

### PR DESCRIPTION
**Pull request by t-vi**
Hi,

this attempts to convert pointwise.py to new-style functions.

I have tried to stay to close to the original and close to the argument order of torch.foo functions (in particular AddCmul and AddCdiv).
Also, I am not 100% sure about the backward in _ConstantGrad - the old implementation had quite an elaborate allocation and I just wrote grad_output.mul(cls.grad_value) without fully understanding the motivation for the old code.

I have converted a few inplace computations to not inplace to deal with gradients being variables.

It seems to pass the test/test_autograd.py and flake8.

Thank you for considering this pull request.

Thomas

### CODE REVIEW

File: torch/autograd/_functions/pointwise.py [hunk 1]
```diff
@@ -138,103 +156,123 @@ def backward(ctx, grad_output):
...
-    def backward(self, grad_output):
-        i, = self.saved_tensors
-        return grad_output * i.mul(i).add_(1).reciprocal_()
+    @staticmethod
+    def backward(ctx, grad_output):
+        i, = ctx.saved_variables
+        return grad_output * i.mul(i).add(1).reciprocal()
 
 
 class Reciprocal(Function):
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: Can you add a TODO saying "make inplace and update grad formulas" here? We could add an inplace version and make many of the definitons you modified more efficient by using `reciprocal_()`

File: torch/autograd/_functions/pointwise.py [hunk 2]
```diff
@@ -138,103 +156,123 @@ def backward(ctx, grad_output):
...
+        ctx.save_for_backward(i)
         return i.atan()
 
-    def backward(self, grad_output):
-        i, = self.saved_tensors
-        return grad_output * i.mul(i).add_(1).reciprocal_()
+    @staticmethod
+    def backward(ctx, grad_output):
+        i, = ctx.saved_variables
+        return grad_output * i.mul(i).add(1).reciprocal()
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: you should be able to use `add_` here

File: torch/autograd/_functions/pointwise.py [hunk 3]
```diff
@@ -138,103 +156,123 @@ def backward(ctx, grad_output):
...
+    def forward(ctx, i):
+        ctx.save_for_backward(i)
         return i.sqrt()
 
-    def backward(self, grad_output):
-        i, = self.saved_tensors
+    @staticmethod
+    def backward(ctx, grad_output):
+        i, = ctx.saved_variables
         return grad_output.mul(i.pow(-0.5)).div(2)
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: It should be possible to use `div_` here

File: torch/autograd/_functions/pointwise.py [hunk 4]
```diff
@@ -138,103 +156,123 @@ def backward(ctx, grad_output):
...
+    @staticmethod
+    def forward(ctx, a, b):
+        ctx.save_for_backward(a, b)
         return a.max(b)
 
-    def backward(self, grad_output):
+    @staticmethod
+    def backward(ctx, grad_output):
+        a, b = ctx.saved_variables
+        mask = a.gt(b).type_as(a)
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: It's better to save the mask in forward (just `a.gt(b)` and convert it to `grad_output`'s type here (it's going to save a lot of memory - if a and b are float tensors it's an 8x reduction!). To increase the efficiency even more you can guard the `ctx.mask = a.gt(b)` with an `if ctx.requires_grad`

File: torch/autograd/_functions/pointwise.py [hunk 5]
```diff
@@ -255,14 +293,18 @@ def backward(ctx, grad_output):
...
-    def backward(self, grad_output):
+    @staticmethod
+    def backward(ctx, grad_output):
+        a, b = ctx.saved_variables
+        mask = a.lt(b).type_as(a)
         return (
-            grad_output * self._min_buffer,
-            grad_output * self._min_buffer.eq(0).type_as(grad_output)
+            grad_output * mask,
+            grad_output * mask.eq(0).type_as(grad_output)
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: As above

**Comment by t-vi** [comment]
O_o I should run the full test suite.

File: torch/autograd/_functions/pointwise.py [hunk 6]
```diff
@@ -284,17 +326,14 @@ def backward(ctx, grad_output):
...
+        return getattr(args[0], cls.__name__.lower())(*args[1:])
 
-    def forward(self, i):
-        return getattr(i, type(self).__name__.lower())(*self.args)
-
-    def backward(self, grad_output):
-        grad_input = grad_output.new(*repeat(1, grad_output.dim()))
-        grad_input = grad_input.fill_(self.grad_value).expand_as(grad_output)
-        return grad_input.mul(grad_output)
+    @classmethod
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: it seems that you're not using `cls` anywhere so it could be a `staticmethod`

File: torch/autograd/_functions/pointwise.py [hunk 7]
```diff
@@ -331,93 +370,133 @@ class Remainder(_ConstantGrad):
...
+    def forward(ctx, a, b, weight):
+        ctx._weight = float(weight)
+        return a.lerp(b, ctx._weight)
 
-    def backward(self, grad_output):
-        return grad_output.mul(1 - self.weight), grad_output.mul(self.weight)
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output.mul(1 - ctx._weight), grad_output.mul(ctx._weight), None
+        # TODO: grad for weight?
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: no need for grad here - weight is a `float`, so autograd can't handle it anyway

File: torch/autograd/_functions/pointwise.py [hunk 8]
```diff
@@ -331,93 +370,133 @@ class Remainder(_ConstantGrad):
...
-        super(Addcmul, self).__init__(inplace)
-        self.scale = scale
-
-    def forward(self, add_tensor, mul_tensor1, mul_tensor2):
-        self.save_for_backward(mul_tensor1, mul_tensor2)
-        if self.inplace:
-            self.mark_dirty(add_tensor)
-            return add_tensor.addcmul_(self.scale, mul_tensor1, mul_tensor2)
+    @staticmethod
+    def forward(ctx, add_tensor, *args, **argsd):
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: No need to capture kwargs -- functions don't support them

File: torch/autograd/_functions/pointwise.py [hunk 9]
```diff
@@ -331,93 +370,133 @@ class Remainder(_ConstantGrad):
...
+        inplace = argsd.get("inplace", False)
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: You need to obtain that from `args`

File: torch/autograd/_functions/pointwise.py [hunk 10]
```diff
@@ -331,93 +370,133 @@ class Remainder(_ConstantGrad):
...
-        super(Addcdiv, self).__init__(inplace)
-        self.scale = scale
-
-    def forward(self, add_tensor, div_tensor1, div_tensor2):
-        self.save_for_backward(div_tensor1, div_tensor2)
-        if self.inplace:
-            self.mark_dirty(add_tensor)
-            return add_tensor.addcdiv_(self.scale, div_tensor1, div_tensor2)
+    @staticmethod
+    def forward(ctx, add_tensor, *args, **argsd):
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: can't you just make the signature `(ctx, add_tensor, div_tensor1, div_tensor2, scale=1)`? This would simplify the code a lot, and it's simple to modify the Variable code to call it like that

File: torch/autograd/_functions/pointwise.py (hunk 4 above)
Comment (t-vi) [inline torch/autograd/_functions/pointwise.py]: Yes, thanks. How do I reliably make a variable from ctx.mask in the backward?

File: torch/autograd/_functions/pointwise.py (hunk 10 above)
Comment (t-vi) [inline torch/autograd/_functions/pointwise.py]: Yes, I'd love to. :)

File: torch/autograd/_functions/pointwise.py (hunk 6 above)
Comment (t-vi) [inline torch/autograd/_functions/pointwise.py]: I think I use cls in both methods  (in forward for the `cls.__name__`, in backward for the `cls.grad_value`).

**Comment by t-vi** [comment]
Thank you for helping me improve my code! I hope this is better.

File: torch/autograd/_functions/pointwise.py (hunk 6 above)
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: My bad, you're right

File: torch/autograd/_functions/pointwise.py [hunk 11]
```diff
@@ -331,93 +362,90 @@ class Remainder(_ConstantGrad):
...
 
-        if self.needs_input_grad[1]:
-            grad_mul1 = grad_output.mul(mul_tensor2).mul(self.scale)
+        if ctx.needs_input_grad[1]:
+            grad_mul1 = grad_output.mul(mul_tensor2).mul(ctx._scale)
 
-        if self.needs_input_grad[2]:
-            grad_mul2 = grad_output.mul(mul_tensor1).mul(self.scale)
+        if ctx.needs_input_grad[2]:
+            grad_mul2 = grad_output.mul(mul_tensor1).mul(ctx._scale)
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: You can do `.mul_(ctx._scale)` here and above (same in cdiv)

File: torch/autograd/_functions/pointwise.py [hunk 12]
```diff
@@ -331,93 +362,90 @@ class Remainder(_ConstantGrad):
...
+            grad_div1 = grad_output.div(div_tensor2).mul(ctx._scale)
 
-        if self.needs_input_grad[2]:
+        if ctx.needs_input_grad[2]:
             div_tensor2_sq = div_tensor2.mul(div_tensor2)
-            grad_div2 = grad_output.mul(div_tensor1).div_(div_tensor2_sq)
-            grad_div2.neg_().mul_(self.scale)
-
-        return grad_add, grad_div1, grad_div2
+            grad_div2 = grad_output.mul(div_tensor1).div(div_tensor2_sq).neg().mul(ctx._scale)
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: you can use `neg_()`

File: torch/autograd/variable.py [hunk 13]
```diff
@@ -582,10 +582,10 @@ def dot(self, other):
...
         if len(args) == 3:
             # scale, tensor1, tensor2
-            return op(args[0])(self, *args[1:])
+            return op.apply(self, args[1], args[2], args[0])
         else:
             # tensor1, tensor2
-            return op()(self, *args)
+            return op.apply(self, *args)
 
     def addcmul(self, *args):
```
Comment (apaszke) [inline torch/autograd/variable.py]: Can you also expose inplace versions of these function? (Just modify `_addcop` to accept a third `inplace` arg)? It seems that we forgot about them. Please also add them to method tests.

**Review by apaszke — COMMENTED** [review]
Looks great! Just 3 minor comments and it's good to go!

File: torch/autograd/_functions/pointwise.py (hunk 12 above)
Comment (fmassa) [inline torch/autograd/_functions/pointwise.py]: So, some in-place ops in Variables can be used without worrying too much? I suppose this only applies to invertible ops, like `neg_`?

**Comment by t-vi** [comment]
Trying to get the CI to try again

**Comment by t-vi** [comment]
So I think I have the comments covered except for the inplace test which I understand is automatic and not manual.

File: torch/autograd/_functions/pointwise.py [hunk 14]
```diff
@@ -331,93 +362,90 @@ class Remainder(_ConstantGrad):
...
+            grad_div1 = grad_output.div(div_tensor2).mul_(ctx._scale)
 
-        if self.needs_input_grad[2]:
+        if ctx.needs_input_grad[2]:
             div_tensor2_sq = div_tensor2.mul(div_tensor2)
-            grad_div2 = grad_output.mul(div_tensor1).div_(div_tensor2_sq)
-            grad_div2.neg_().mul_(self.scale)
-
-        return grad_add, grad_div1, grad_div2
+            grad_div2 = grad_output.mul(div_tensor1).div(div_tensor2_sq).neg_().mul_(ctx._scale)
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: Just one last change. Can you replace `.neg_().mul_(ctx._scale)` with `.mul(-ctx._scale)`?

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Can you please also fix the conflicts?

**Comment by soumith** [comment]
I resolved conflicts and made that change.

**Review by apaszke — APPROVED** [review]

**Comment by t-vi** [comment]
Hi

Apologies, I've been travelling with no access to my torch computer. Thanks for fixing my patches.
Best regards

Thomas
